File: gpslogger/src/main/java/com/mendhak/gpslogger/loggers/customurl/CustomUrlRequest.java
package com.mendhak.gpslogger.loggers.customurl;

import android.support.v4.util.Pair;

import com.mendhak.gpslogger.common.Strings;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class CustomUrlRequest implements Serializable {


    private HashMap<String, String> httpHeaders = new HashMap<String, String>();
    private String logURL;
    private String httpMethod;
    private String httpBody;
    private String rawHeaders;



    public CustomUrlRequest(String logUrl, String httpMethod) {
        this(logUrl, httpMethod, "", "");
    }

    public CustomUrlRequest(String logUrl)  {
        this(logUrl, "GET");
    }

    public CustomUrlRequest(String logURL, String httpMethod, String httpBody, String rawHeaders){
        this.logURL = logURL;
        this.httpMethod = httpMethod.toUpperCase();
        this.httpBody = httpBody;
        this.rawHeaders = rawHeaders;


        Pair<String, String> creds = getBasicAuthCredentialsFromUrl(this.logURL);
        addAuthorizationHeader(creds);
        removeCredentialsFromUrl(creds);

        //HttpHeaders.putAll(getHeadersFromTextBlock(RawHeaders))
        this.httpHeaders.putAll(getHeadersFromTextBlock(rawHeaders));


    }

    private Map<String,String> getHeadersFromTextBlock(String rawHeaders) {

        HashMap<String, String> map = new HashMap<>();
        String[] lines = rawHeaders.split("\\r?\\n");
        for (String line : lines){
            if(!Strings.isNullOrEmpty(line) && line.contains(":")){
                String[] lineParts = line.split(":");
                if(lineParts.length == 2){
                    String lineKey = line.split(":")[0].trim();
                    String lineValue = line.split(":")[1].trim();

                    if(!Strings.isNullOrEmpty(lineKey) && !Strings.isNullOrEmpty(lineValue)){
                        map.put(lineKey, lineValue);
                    }
                }

            }
        }

        return map;

    }

    private void addAuthorizationHeader(Pair<String, String> creds) {

        if(!Strings.isNullOrEmpty(creds.first) && !Strings.isNullOrEmpty(creds.second)){
            String credential = okhttp3.Credentials.basic(creds.first, creds.second);
            this.httpHeaders.put("Authorization", credential);
        }

    }

    private void removeCredentialsFromUrl(Pair<String, String> creds) {
        this.logURL = this.logURL.replace(creds.first + ":" + creds.second + "@","");
    }

    private Pair<String, String> getBasicAuthCredentialsFromUrl(String logURL) {
        Pair<String, String> result  = new Pair<>("","");

        Pattern r = Pattern.compile("(\\w+):(\\w+)@.+");
        Matcher m = r.matcher(logURL);
        while(m.find()){
            result = new Pair<>(m.group(1), m.group(2));
        }

        return result;

    }



    String getHttpMethod(){
        return this.httpMethod;
    }

    String getLogURL(){
        return this.logURL;
    }

    HashMap<String, String> getHttpHeaders(){
        return this.httpHeaders;
    }

    String getHttpBody(){
        return this.httpBody;
    }
}

File: gpslogger/src/main/java/com/mendhak/gpslogger/loggers/customurl/CustomUrlRequest.kt
package com.mendhak.gpslogger.loggers.customurl

import java.util.HashMap
import java.util.regex.Pattern
import okhttp3.Credentials


data class CustomUrlRequest @JvmOverloads constructor (var LogURL : String,
                                                       var HttpMethod: String = "GET",
                                                       var HttpBody : String = "",
                                                       var RawHeaders : String = "") : java.io.Serializable  {


    var HttpHeaders : HashMap<String,String> = hashMapOf()


    init {
        HttpMethod = HttpMethod.toUpperCase()
    }

    init {
        val (usr,pwd) = getBasicAuthCredentialsFromUrl(LogURL)

        addAuthorizationHeader(usr,pwd)
        removeCredentialsFromUrl(usr,pwd)
    }

    init {
        HttpHeaders.putAll(getHeadersFromTextBlock(RawHeaders))
    }

    private fun removeCredentialsFromUrl(basicAuthUsername: String, basicAuthPassword: String) {
        LogURL = LogURL.replace("$basicAuthUsername:$basicAuthPassword@","")
    }

    private fun addAuthorizationHeader(usr: String, pwd: String) {
        if(!usr.isBlank() && !pwd.isBlank()){
            val credential = Credentials.basic(usr, pwd)
            HttpHeaders.put("Authorization", credential)
        }
    }

    private fun getBasicAuthCredentialsFromUrl(logUrl: String): Pair<String, String> {

        var result  = Pair<String,String>("","")

        val r = Pattern.compile("(\\w+):(\\w+)@.+")
        val m = r.matcher(logUrl)
        while(m.find()){
            result = Pair<String,String>(m.group(1), m.group(2))
        }

        return result;
    }

    private fun getHeadersFromTextBlock(httpHeaders : String) : HashMap<String,String>{
        val map = hashMapOf<String,String>()
        val lines = httpHeaders.split(regex="\\r?\\n".toRegex())
        for (line in lines){
            if(!line.isBlank() && line.contains(":")){
                val lineKey = line.split(":")[0].trim()
                val lineValue = line.split(":")[1].trim()
                if(!lineKey.isBlank() && !lineKey.isBlank()){
                    map.put(lineKey,lineValue)
                }
            }
        }
        return map
    }

}